### PR TITLE
plan: add a Cache plan in physical plan

### PR DESCRIPTION
**Pull request by winoros**
Currently, sql such as `select id from t2 where exists(select * from t1 where t1.v < t2.v union select * from t3 where t3.a > 1);` will do the subquery `select * from t3 where t3.a > 1` mulitple times which is a waste of time. So we add a Cache plan above this subquery to cache its result.
PTAL @shenli @coocood @zimulala @XuHuaiyu @tiancaiamao @hanfei1991

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -665,3 +667,11 @@ func (b *executorBuilder) buildDelete(v *plan.Delete) Executor {
 		IsMultiTable: v.IsMultiTable,
 	}
 }
+
+func (b *executorBuilder) buildTempStore(v *plan.TempStore) Executor {
+	selExec := b.build(v.GetChildByIndex(0))
```
Comment (hanfei1991) [inline executor/builder.go]: selExec -> src

File: plan/physical_plan_builder.go [hunk 2]
```diff
@@ -668,6 +689,7 @@ func (p *Aggregation) convert2PhysicalPlanStream(prop *requiredProperty) (*physi
 		AggFuncs:     p.AggFuncs,
 		GroupByItems: p.GroupByItems,
 	}
+	agg.correlated = p.IsCorrelated()
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: Why add this line ?

File: executor/executor.go [hunk 3]
```diff
@@ -1931,3 +1931,36 @@ func (e *DummyScanExec) Close() error {
...
+	cursor     int
+}
+
+// Schema implements the Executor Schema interface.
+func (e *TempStoreExec) Schema() expression.Schema {
+	return e.schema
+}
+
+// Close implements the Executor Close interface.
+func (e *TempStoreExec) Close() error {
```
Comment (hanfei1991) [inline executor/executor.go]: e.Src.Close() to release its src's resource.

File: plan/match_property.go [hunk 4]
```diff
@@ -252,6 +252,13 @@ func (p *Projection) matchProperty(_ *requiredProperty, childPlanInfo ...*physic
 }
 
 // matchProperty implements PhysicalPlan matchProperty interface.
+func (p *TempStore) matchProperty(prop *requiredProperty, childPlanInfo ...*physicalPlanInfo) *physicalPlanInfo {
+	np := *p
```
Comment (hanfei1991) [inline plan/match_property.go]: you don't need to implent this function. Just "panic("You can't call this function!")"

File: plan/physical_plans.go [hunk 5]
```diff
@@ -759,3 +764,20 @@ func (p *PhysicalUnionScan) Copy() PhysicalPlan {
...
 }
+
+// Copy implements the PhysicalPlan Copy interface.
+func (p *TempStore) Copy() PhysicalPlan {
+	np := *p
+	return &np
+}
+
+// MarshalJSON implements json.Marshaler interface.
+func (p *TempStore) MarshalJSON() ([]byte, error) {
```
Comment (hanfei1991) [inline plan/physical_plans.go]: You don't need to implement it.

Comment (ngaut) [inline plan/physical_plans.go]: The go style is "panic("unimplemented") "

File: plan/physical_plan_builder.go (hunk 2 above)
Comment (winoros) [inline plan/physical_plan_builder.go]: if don't add this. the value of correlated of one PhysicalAggregation will always be false though the correlated of its logical plan can be true.

File: executor/executor.go [hunk 6]
```diff
@@ -1931,3 +1931,41 @@ func (e *DummyScanExec) Close() error {
...
+	if !e.closed {
+		e.closed = true
+		return e.Src.Close()
+	}
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *TempStoreExec) Next() (*Row, error) {
+	if e.cursor == len(e.storedRows) {
```
Comment (hanfei1991) [inline executor/executor.go]: This part makes me confused. You can write like this : if not closed, prepare the storedRows and return srcRow, if the row from src is nil, set closed true and close the src. If closed is true, use the increasing cursor to read data. If cursor is out of index, return nil.

**Comment by hanfei1991** [comment]
fix ci @winoros

File: executor/executor.go [hunk 7]
```diff
@@ -1931,3 +1931,49 @@ func (e *DummyScanExec) Close() error {
...
+// Close implements the Executor Close interface.
+func (e *TempStoreExec) Close() error {
+	e.cursor = 0
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *TempStoreExec) Next() (*Row, error) {
+	if e.closed {
+		e.cursor++
```
Comment (coocood) [inline executor/executor.go]: In all other executors, we update cursor after getting a row, we better keep it consistent.

File: executor/executor.go [hunk 8]
```diff
@@ -1931,3 +1931,49 @@ func (e *DummyScanExec) Close() error {
 func (e *DummyScanExec) Next() (*Row, error) {
 	return nil, nil
 }
+
+// TempStoreExec represents TempStore executor.
+// it stores the return values of the executor of its child node.
+type TempStoreExec struct {
```
Comment (coocood) [inline executor/executor.go]: I think `CacheExec` can better describe what this executor does.

**Comment by hanfei1991** [comment]
LGTM

File: executor/executor.go [hunk 9]
```diff
@@ -1932,48 +1932,46 @@ func (e *DummyScanExec) Next() (*Row, error) {
...
+			e.srcFinished = true
 			err := e.Src.Close()
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
 		}
-		return row, nil
+		e.storedRows = append(e.storedRows, row)
 	}
+	e.cursor++
```
Comment (coocood) [inline executor/executor.go]:
```
row := e.storedRows[e.cursor]
e.cursor++
return row, nil
```

File: executor/executor.go [hunk 10]
```diff
@@ -1931,3 +1931,48 @@ func (e *DummyScanExec) Close() error {
...
+}
+
+// Schema implements the Executor Schema interface.
+func (e *CacheExec) Schema() expression.Schema {
+	return e.schema
+}
+
+// Close implements the Executor Close interface.
+func (e *CacheExec) Close() error {
+	e.cursor = 0
```
Comment (coocood) [inline executor/executor.go]: Call `e.Src.Close` is common pattern.

**Comment by coocood** [comment]
LGTM

**Pull request merged**